File: scala/leetcode/src/main/scala/leetcode/editor/cn/[25]K 个一组翻转链表.scala
package leetcode.editor.cn
object ReverseNodesInKGroup {
//leetcode submit region begin(Prohibit modification and deletion)
  /** Definition for singly-linked list. class ListNode(_x: Int = 0, _next: ListNode = null) { var next: ListNode = _next var x: Int = _x }
    */
  object Solution {
    def reverseKGroup(head: ListNode, k: Int): ListNode =
      val dummy = new ListNode(0, head)
      (0 until nodeCount(head) / k).foldLeft((dummy, head)) { case ((prev, head), _) =>
        val (newHead, newTail) = reverseFirstN(head, k)
        prev.next = newHead
        head.next = newTail
        (head, newTail)
      }

      dummy.next

    @annotation.tailrec
    private def nodeCount(node: ListNode, acc: Int = 0): Int =
      if node == null then acc
      else nodeCount(node.next, acc + 1)

    @annotation.tailrec
    private def reverseFirstN(head: ListNode, n: Int, acc: ListNode = null): (ListNode, ListNode) =
      if n == 0 || head == null
      then (acc, head)
      else
        val next = head.next
        head.next = acc
        reverseFirstN(next, n - 1, head)
  }
//leetcode submit region end(Prohibit modification and deletion)

}

@main
def run(): Unit =
  println(ReverseNodesInKGroup.Solution.reverseKGroup(new ListNode(1, new ListNode(2, null)), 2))

File: cpp/leetcodecn/[25]K 个一组翻转链表.cpp
#include <vector>
#include "util.h"
#include <map>
#include <string>
#include <utility>
#include <memory>
#include <ranges>
#include <algorithm>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *reverseKGroup(ListNode *head, int k) {
        if (head == nullptr) return nullptr;

        auto dummy = make_unique<ListNode>(-1, head);
        auto curr = head;
        auto prev = dummy.get();
        auto count = nodeCount(head);
        for (auto i = 0; i < count / k; i++) {
            auto [newHead, newTail] = reserveN(curr, k);
            prev->next = newHead;
            curr->next = newTail;
            prev = curr;
            curr = newTail;
        }

        return dummy->next;
    }

    int nodeCount(ListNode *node, size_t acc = 0) {
        if (node == nullptr) return acc;
        else return nodeCount(node->next, acc + 1);
    }

    pair<ListNode *, ListNode *> reserveN(ListNode *head, int n, ListNode *acc = nullptr) {
        if (head == nullptr || n == 0)
            return make_pair(acc, head);
        else {
            auto next = head->next;
            head->next = acc;
            return reserveN(next, n - 1, head);
        }
    }
};
//leetcode submit region end(Prohibit modification and deletion)


int main(int argv, char *argc[]) {
}